src/jit/flowgraph.cpp

```diff
@@ -5746,14 +5746,47 @@ void          Compiler::fgFindBasicBlocks()
 
     if (compIsForInlining())
     {
+        bool hasReturnBlocks = false;
+        bool hasMoreThanOneReturnBlock = false;
+
+        for (BasicBlock* block = fgFirstBB; block != nullptr; block = block->bbNext)
+        {
+            if (block->bbJumpKind == BBJ_RETURN)
+            {
+                if (hasReturnBlocks)
+                {
+                    hasMoreThanOneReturnBlock = true;
+                    break;
+                }
+
+                hasReturnBlocks = true;
+            }
+        }
+
+        if (!hasReturnBlocks)
+        {
+            //
+            // Mark the call node as "no return". The inliner might ignore CALLEE_DOES_NOT_RETURN and
+            // fail inline for a different reasons. In that case we still want to make the "no return"
+            // information available to the caller as it can impact caller's code quality.
+            //
+
+            impInlineInfo->iciCall->gtCallMoreFlags |= GTF_CALL_M_DOES_NOT_RETURN;
+
+            compInlineResult->NoteInt(InlineObservation::CALLEE_DOES_NOT_RETURN, static_cast<int>(fgBBcount));
```

Comment:
I'd really like this to be a `NoteBool` so that this observation is policy neutral. That creates a complication on the policy end, since any rejection will now be the result of two observations, and (at least from the policy's viewpoint) there's no guarantee on observation order. What I'd suggest is that for the earlier observation, we add a DEBUG mode  "we have seen this observation" flag and assert that it's true when the later observation is made.